gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+                in.nextNull();
+                return null;
+            }
+            else if (in.peek() != JsonToken.BEGIN_OBJECT) {
+                throw new DeSerializeException("Expecting null or begin object at [" + in.getPath() + "]");
+            }
+            in.beginObject();
+            // This array will hold the value for each json field in the src object that we find, deserialized
+            // using a RecordField, that will also respect all Gson annotations. For fields that are not deserialized,
+            // the default null value will be used.
+            Object[] values = new Object[recordFields.length];
+            nextField:
+            while (in.peek() != JsonToken.END_OBJECT) {
+                String fieldName = in.nextName();
+                // Check if we know which field this name should deserialize to. If we do not find a matching index,
+                // then this field is one that is not present in our record, and ignored. We do not treat unknown
+                // object fields as an error, because the version of the class we have might differ from the source.
+                // -stun.
+                Integer fieldIndex = recordFieldIndex.get(fieldName);
+                if (fieldIndex != null) {
+                    values[fieldIndex] = recordFields[fieldIndex].read(in);
+                }
+                else {
+                    // Be sure to skip values we do not consume, otherwise the state of the parser will be wrong.
+                    in.skipValue();
+                }
+            }
+            in.endObject();
+            // At this point, we have not verified that all fields are present, we treat absent values as null
+            // before passing on to the constructor. We could verify that all fields are present, but the src could
+            // have different rules about serializing null values than we do. So it is better to ignore that we are
+            // missing some values, rather than throw here. -stun.
+            try {
+                return constructor.newInstance(values);
+            }
+            catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+                throw new DeSerializeAccessException(
+                        "Failed to create an instance of [" + constructor.getDeclaringClass() + "] at [" + in.getPath()
+                        + "]", e);
+            }
+        }
+    }
+
+    // ========= Exceptions ============================================================================================
+
+    public static class TypeAdapterReflectionException extends RuntimeException {
+
+        private static final long serialVersionUID = -3905048094470359103L;
+
+        public TypeAdapterReflectionException(String message, Throwable cause) {
+            super(message, cause);
+        }
+    }
+
+    public static class DeSerializeAccessException extends RuntimeException {
```

Comment:
`DeserializeAccessException`, and `DeserializeException` below.

Reply:
Fixed.